docs/sts/tls.md

```diff
@@ -0,0 +1,126 @@
+# AssumeRoleWithCertificate [![Slack](https://slack.min.io/slack?type=svg)](https://slack.min.io)
+
+## Introduction
+
+MinIO provides a custom STS API that allows authentication with client X.509 / TLS certificates.
+
+A major advantage of certificate-based authentication compared to other STS authentication methods,
+like OpenID Connect or LDAP/AD, is that client authentication works without any additional/external
+component that must be constantly available. Therefore, certificate-based authentication may provide
+better availability / lower operational complexity.
+
+The MinIO TLS STS API can be configured via MinIO's standard configuration API (i.e. using `mc admin config set/get`).
+Further, it can be configured via the following environment variables:
+
+```
+mc admin config set myminio identity_tls --env
+KEY:
+identity_tls  enable X.509 TLS certificate SSO support
+
+ARGS:
+MINIO_IDENTITY_TLS_STS_EXPIRY   (duration)  temporary credentials validity duration in s,m,h,d. Default is "1h"
+MINIO_IDENTITY_TLS_SKIP_VERIFY  (on|off)    trust client certificates without verification. Defaults to "off" (verify)
+```
+
+The MinIO TLS STS API is enabled by default. However, it can be completely *disabled* by setting:
+```
+MINIO_IDENTITY_TLS_ENABLE=off
+```
+
+## Example
+
+MinIO exposes a custom S3 STS API endpoint as `Action=AssumeRoleWithCertificate`. A client has to send an HTTP `POST`
+request to `https://<host>:<port>?Action=AssumeRoleWithCertificate&Version=2011-06-15`. Since the authentication and
+authorization happens via X.509 certificates the client has to send the request over **TLS** and has to provide
+a client certificate.
+
+The following curl example shows how to authenticate to a MinIO server with client certificate and obtain temp. S3
+access/secret key credentials. 
+
+```curl
+curl -X POST --key private.key --cert public.crt "https://minio:9000?Action=AssumeRoleWithCertificate&Version=2011-06-15"
+```
+
+```xml
+<?xml version="1.0" encoding="UTF-8"?>
+<AssumeRoleWithCertificateResponse xmlns="https://sts.amazonaws.com/doc/2011-06-15/">
+   <AssumeRoleWithCertificateResult>
+      <Credentials>
+         <AccessKeyId>YC12ZBHUVW588BQAE5BM</AccessKeyId>
+         <SecretAccessKey>Zgl9+zdE0pZ88+hLqtfh0ocLN+WQTJixHouCkZkW</SecretAccessKey>
+         <Expiration>2021-07-19T20:10:45Z</Expiration
+         <SessionToken>eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3NLZXkiOiJZQzEyWkJIVVZXNTg4QlFBRTVCTSIsImV4cCI6MTYyNjcyNTQ0NX0.wvMUf3w_x16qpVWgua8WxnV1Sgtv1jOnSu03vbrwOMzV3cI4q3_9WZD9LwlP-34DTsvbsg7gCBGh6YNriMMiQw</SessionToken>
+      </Credentials>
+   </AssumeRoleWithCertificateResult>
+   <ResponseMetadata>
+      <RequestId>169339CD8B3A6948</RequestId>
+   </ResponseMetadata>
+</AssumeRoleWithCertificateResponse>
+```
+
+## Authentication Flow
```

Comment:
@aead how does this work with arbitrarily signed x.509 certificates? That is, does this API require that the client specify an x.509 cert signed by a known CA (CA is in ~/.minio/certs/CAs) ?

I'm specifically thinking of a scenario where an arbitrary client uses this API with a self-signed certificate with `CN = consoleAdmin` to gain admin access to the MinIO cluster.

I think most users would want to block this API to only those certificates signed by a trusted CA (at minimum). From there I might want to go farther and only trust those certificates whose CN matches some group filter, e.g.:

```
MINIO_IDENTITY_TLS_DN_FILTER="[OU=OrgUnit], [O=ORG], [L=City],[S=STATE],[C=COUNTRYCODE]
```

This would ensure that only those x.509 certs whose DN includes that specific string pattern after the CN would work.

So if I had:

```
MINIO_IDENTITY_TLS_DN_FILTER="OU=datalake,O=techops"
```

I would only expect the first of the following three x.509 certs to work with this API:

```
Subject: CN = ConsoleAdmin, OU=datalake,O=techops
Subject: CN = ConsoleAdmin, O=techops
Subject: CN = ConsoleAdmin
```

Let me know if this makes sense. This is a really powerful API, but since it is enabled by default IMHO we need some guardrails to prevent "guessing" access to roles by anyone with cert-minting capabilities.

Reply:
This is something that we can look at it

Reply:
I'll try testing this week. I'd like to at least confirm we restrict to trusted CA.

Reply:
mTLS won't work without it @ravindk89

Reply:
Yeah, it's more for my own curiosity as well. Easier to document that way too.